neps/nep-0364.md

```diff
@@ -146,6 +146,15 @@ This NEP aims to introduce the following host function:
 ///
 /// `input_cost(num_bytes_signature + num_bytes_message, num_bytes_public_key) +
 ///  ed25519_verify_base + ed25519_verify_byte * (num_bytes_signature + num_bytes_message)`
+///
+/// # Errors
+///
+/// The signature size is fixed and equal to SIGNATURE_LENGTH (64 bytes). In case the length of signature input is not equal to
+/// SIGNATURE_LENGTH, then the function returns HostError::Ed25519VerifyInvalidInput with the message "invalid signature length".
+/// 
+/// A similar case occurs with the public key. Its size is known and its equal to PUBLIC_KEY_LENGTH (32 bytes). In case the
+/// input length provided for the publc key doesn't match PUBLIC_KEY_LENGTH, the function will return the following error:
+/// HostError::Ed25519VerifyInvalidInput with the message "invalid public key length".
```

Comment:
@matklad Do you think it is a good direction to go with extending `HostError` type or should we keep the scope of potential errors minimal?

Reply:
I think this doesn't matter much: from the protocol perspective, how exactly `HostError` looks is unobservable. Implementation wise, adding a new variant there is OK.


But this actually points to the problem with wording of this NEP: what is written here is how the function would look inside `logic.rs`, but that's an impl detail which is completely irrelevant from the specification point of view. 

NEPs should say how the WASM import looks, not how it is implemented. For the time being, I'd rephrase this section as:

== Specification

The following host function is added

```rust
extern "C"{

/// Verify an ED25519 signature given a message and a public key.
/// # Returns
/// - 1 if the signature was properly verified
/// - 0 if the signature failed to be verified
///
/// # Cost
///
/// Each input can either be in memory or in a register. Set the length of the input to `u64::MAX`
/// to declare that the input is a register number and not a pointer.
/// Each input has a gas cost input_cost(num_bytes) that depends on whether it is from memory
/// or from a register. It is either read_memory_base + num_bytes * read_memory_byte in the
/// former case or read_register_base + num_bytes * read_register_byte in the latter. This function
/// is labeled as `input_cost` below.
///
/// `input_cost(num_bytes_signature + num_bytes_message + num_bytes_public_key) +
///  ed25519_verify_base + ed25519_verify_byte * num_bytes_message`
///
/// # Errors
///
/// If the signature size is not equal to 64 bytes, or public key length is not equal to 32 bytes, contract execution is terminated with an error. 
  fn ed25519_verify(
    sig_len: u64,
    sig_ptr: u64,
    msg_len: u64,
    msg_ptr: u64,
    pub_key_len: u64,
    pub_key_ptr: u64,
  ) -> u64;
}
```

That is: 

* write the exten function from perspecive of the contract
* don't mention HostError, as that is unobservable to contracts (it is more or less like panic)

We could perhaps consider using proper wasm for the spec here

```wasm
(module 
  (import "env" "ed25519_verify" (func (param 64) ... (result i64)))
)
```

but I feel that Rust-syntax to experss that is more accessible, and better matches what we've been doing historically

Reply:
Updated it